homeassistant/components/asuswrt/bridge.py

```diff
@@ -271,3 +322,163 @@ async def _get_temperatures(self) -> dict[str, Any]:
+        except AsusWrtError as exc:
+            raise UpdateFailed(exc) from exc
+        return {
+            format_mac(mac): WrtDevice(dev.ip, dev.name, dev.node)
+            for mac, dev in api_devices.items()
+        }
+
+    async def async_get_available_sensors(self) -> dict[str, dict[str, Any]]:
+        """Return a dictionary of available sensors for this bridge."""
+        sensors_temperatures = await self._get_available_temperature_sensors()
+        sensors_cpu = await self._get_available_cpu_sensors()
+        sensors_types = {
+            SENSORS_TYPE_BYTES: {
+                KEY_SENSORS: SENSORS_BYTES,
+                KEY_METHOD: self._get_bytes,
+            },
+            SENSORS_TYPE_CPU: {
+                KEY_SENSORS: sensors_cpu,
+                KEY_METHOD: self._get_cpu_usage,
+            },
+            SENSORS_TYPE_MEMORY: {
+                KEY_SENSORS: SENSORS_MEMORY,
+                KEY_METHOD: self._get_memory_usage,
+            },
+            SENSORS_TYPE_RATES: {
+                KEY_SENSORS: SENSORS_RATES,
+                KEY_METHOD: self._get_rates,
+            },
+            SENSORS_TYPE_TEMPERATURES: {
+                KEY_SENSORS: sensors_temperatures,
+                KEY_METHOD: self._get_temperatures,
+            },
+        }
+        return sensors_types
+
+    async def _get_available_cpu_sensors(self) -> list[str]:
+        """Check which cpu information is available on the router."""
+        try:
+            available_cpu = await self._api.async_get_cpu_usage()
+            available_sensors = [t for t in SENSORS_CPU if t in available_cpu]
+        except AsusWrtError as exc:
+            _LOGGER.warning(
+                (
+                    "Failed checking cpu sensor availability for ASUS router"
+                    " %s. Exception: %s"
+                ),
+                self.host,
+                exc,
+            )
+            return []
+        return available_sensors
+
+    async def _get_available_temperature_sensors(self) -> list[str]:
+        """Check which temperature information is available on the router."""
+        try:
+            available_temps = await self._api.async_get_temperatures()
+            available_sensors = [
+                t for t in SENSORS_TEMPERATURES if t in available_temps
+            ]
+        except AsusWrtError as exc:
+            _LOGGER.warning(
+                (
+                    "Failed checking temperature sensor availability for ASUS router"
+                    " %s. Exception: %s"
+                ),
+                self.host,
+                exc,
+            )
+            return []
+        return available_sensors
+
+    async def _get_bytes(self) -> dict[str, Any]:
+        """Fetch byte information from the router."""
+        try:
+            datas = await self._api.async_get_traffic_bytes()
+        except AsusWrtError as exc:
+            raise UpdateFailed(exc) from exc
```

Comment:
This try/except pattern is exactly the same every single time. This could use a decorator.

And doing that, will make it easier to just use it in your coordinator directly, without the need to wrap every API call in a new proxy method.

Reply:
Not really clear, especially the second part. Non all proxy method will be identical. But may be we can work on this on next PR when I will add additional sensors?

Reply:
You could write a decorator who does this steps: (not a complete decorator example)
```python
try:
    data = await func(self, *args, **kwargs)
except exceptions as exc:
    raise UpdateFailed(exc) from exc

return _get_dict(keys, list(data.values()))
```

and than modify the function like this:
```python
    @handle_errors_and_zip(AsusWrtError, SENSORS_BYTES)
    async def _get_bytes(self) -> dict[str, Any]:
        """Fetch byte information from the router."""
        return await self._api.async_get_traffic_bytes()
```